File: src/33.search-in-rotated-sorted-array/33.search-in-rotated-sorted-array.py
#
# @lc app=leetcode id=33 lang=python3
#
# [33] Search in Rotated Sorted Array
#
# Suppose an array sorted in ascending order is rotated at some pivot unknown to you beforehand.

# (i.e., [0,1,2,4,5,6,7] might become [4,5,6,7,0,1,2]).

# You are given a target value to search. If found in the array return its index, otherwise return -1.

# You may assume no duplicate exists in the array.

# Your algorithm's runtime complexity must be in the order of O(log n).

# Example 1:

# Input: nums = [4,5,6,7,0,1,2], target = 0
# Output: 4
# Example 2:

# Input: nums = [4,5,6,7,0,1,2], target = 3
# Output: -1


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        # Accepted
        # 196/196 cases passed (44 ms)
        # Your runtime beats 94.7 % of python3 submissions
        # Your memory usage beats 99.57 % of python3 submissions (12.9 MB)
        l = len(nums)
        if l < 1:
            return -1
        elif l < 2:
            return 0 if nums[0] == target else -1
        num, n = self.mysort(nums, l)
        left, right = 0, l-1
        result = -1
        while left <= right:
            mid = (left+right)//2
            if num[mid] == target:
                result = mid+n
                if result > l-1:
                    if n != 0:
                        return result-l
                return result
            if num[mid] < target:
                left = mid+1
            else:
                right = mid-1
        return result

    def mysort(self, nums, length):
        head = nums[0]
        for i in range(1, length):
            if nums[i] < head:
                return nums[i:]+nums[:i], i
        return nums, 0

    def search2(self, nums: List[int], target: int) -> int:
        # Accepted
        # 196/196 cases passed (44 ms)
        # Your runtime beats 94.7 % of python3 submissions
        # Your memory usage beats 99.57 % of python3 submissions (12.8 MB)
        try:
            return nums.index(target)
        except:
            return -1

File: src/33.search-in-rotated-sorted-array/33.search-in-rotated-sorted-array.c
/*
 * @lc app=leetcode.cn id=33 lang=c
 *
 * [33] Search in Rotated Sorted Array
 */
// Suppose an array sorted in ascending order is rotated at some pivot unknown to you beforehand.

// (i.e., [0,1,2,4,5,6,7] might become [4,5,6,7,0,1,2]).

// You are given a target value to search. If found in the array return its index, otherwise return -1.

// You may assume no duplicate exists in the array.

// Your algorithm's runtime complexity must be in the order of O(log n).

// Example 1:

// Input: nums = [4,5,6,7,0,1,2], target = 0
// Output: 4
// Example 2:

// Input: nums = [4,5,6,7,0,1,2], target = 3
// Output: -1
// @lc code=start

int search(int *nums, int numsSize, int target)
{
    // Accepted
    // 196/196 cases passed (4 ms)
    // Your runtime beats 90.29 % of c submissions
    // Your memory usage beats 67.76 % of c submissions (7.3 MB)
    if (numsSize < 1)
    {
        return -1;
    }
    else if (numsSize < 2)
    {
        return nums[0] == target ? 0 : -1;
    }

    int i = 0;
    for (i = 0; i < numsSize; i++)
    {
        if (nums[i] < nums[0])
            break;
    }
    merge_sort(nums,numsSize);
    int result = -1, left = 0, right = numsSize - 1;
    while(left<=right){
        int mid = (left+right)>>1;
        if(nums[mid]==target){
            result = mid+i;
            if(result>numsSize-1 && i!=0){
                return result-numsSize;
            }
            return result;
        }
        if(nums[mid]<target){
            left= mid+1;
        }else{
            right = mid-1;
        }
    }
    return result;
}


int min(int x, int y) {
    return x < y ? x : y;
}
void merge_sort(int arr[], int len) {
    int* a = arr;
    int* b = (int*) malloc(len * sizeof(int));
    int seg, start;
    for (seg = 1; seg < len; seg += seg) {
        for (start = 0; start < len; start += seg + seg) {
            int low = start, mid = min(start + seg, len), high = min(start + seg + seg, len);
            int k = low;
            int start1 = low, end1 = mid;
            int start2 = mid, end2 = high;
            while (start1 < end1 && start2 < end2)
                b[k++] = a[start1] < a[start2] ? a[start1++] : a[start2++];
            while (start1 < end1)
                b[k++] = a[start1++];
            while (start2 < end2)
                b[k++] = a[start2++];
        }
        int* temp = a;
        a = b;
        b = temp;
    }
    if (a != arr) {
        int i;
        for (i = 0; i < len; i++)
            b[i] = a[i];
        b = a;
    }
    free(b);
}

// @lc code=end
